### PR TITLE
create Translator for Central and Eastern European Online Library Journals

### PR DESCRIPTION
**Pull request by socheres**
…rnals

### CODE REVIEW

File: Central and Eastern European Online Library Journals.js [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+	GNU Affero General Public License for more details.
+
+	You should have received a copy of the GNU Affero General Public License
+	along with Zotero. If not, see <http://www.gnu.org/licenses/>.
+
+	***** END LICENSE BLOCK *****
+*/
+
+function detectWeb(doc, url) {
+	if (url.includes('article')) return "journalArticle";
```
Comment (AbeJellinek) [inline Central and Eastern European Online Library Journals.js:39]: Could the URL fragment you're matching on just be `/article-detail?`? `article` alone will yield a lot of false positives.

File: Central and Eastern European Online Library Journals.js [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+	});
+	translator.translate();
+}
+
+//scraping abstractNote from HTML, that is not included in Embedded Metadata
+function postProcess(doc, item) {
+	let abstractEntry = ZU.xpathText(doc, '//p[@class="summary"]');
+	if (!item.abstractNote && abstractEntry) item.abstractNote = abstractEntry;
+	item.complete();
+}/** BEGIN TEST CASES **/
```
Comment (AbeJellinek) [inline Central and Eastern European Online Library Journals.js:85]: Add two newlines before `/** BEGIN TEST CASES **/`

File: Central and Eastern European Online Library Journals.js [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+function detectWeb(doc, url) {
+	if (url.includes('article')) return "journalArticle";
+	else if (getSearchResults(doc, true)) return "multiple";
+	else return false;
+}
+
+function getSearchResults(doc, checkOnly) {
+	var items = {};
+	var found = false;
+	var rows = doc.querySelectorAll('a.ng-binding');//Z.debug(rows)//a.ng-binding
```
Comment (AbeJellinek) [inline Central and Eastern European Online Library Journals.js:47]: This doesn't seem to be the right selector. `.description a` should work.

File: Central and Eastern European Online Library Journals.js [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+	translator.translate();
+}
+
+//scraping abstractNote from HTML, that is not included in Embedded Metadata
+function postProcess(doc, item) {
+	let abstractEntry = ZU.xpathText(doc, '//p[@class="summary"]');
+	if (!item.abstractNote && abstractEntry) item.abstractNote = abstractEntry;
+	item.complete();
+}/** BEGIN TEST CASES **/
+var testCases = [
```
Comment (AbeJellinek) [inline Central and Eastern European Online Library Journals.js:86]: Add a test for multiples.

**Review by AbeJellinek — CHANGES_REQUESTED** [review]
Hey, thanks! This looks good. Left a few comments.

File: Central and Eastern European Online Library Journals.js (hunk 1 above)
Comment (socheres) [reply Central and Eastern European Online Library Journals.js]: Thanks for code reviewing and your suggestion. I have changed it.

File: Central and Eastern European Online Library Journals.js (hunk 2 above)
Comment (socheres) [reply Central and Eastern European Online Library Journals.js]: ok!

File: Central and Eastern European Online Library Journals.js (hunk 3 above)
Comment (socheres) [reply Central and Eastern European Online Library Journals.js]: with `.description a` it cannot detect multiple downloads, therefore I have left it as it is.

File: Central and Eastern European Online Library Journals.js (hunk 4 above)
Comment (socheres) [reply Central and Eastern European Online Library Journals.js:89]: done.

File: Central and Eastern European Online Library Journals.js [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+				],
+				"notes": [],
+				"seeAlso": []
+			}
+		]
+	},
+	{
+		"type": "web",
+		"url": "https://www.ceeol.com/search/journal-detail?id=1266",
+		"items": [
```
Comment (AbeJellinek) [inline Central and Eastern European Online Library Journals.js:313]: Should this be a `"multiple"`? It's definitely not a single article.

File: Central and Eastern European Online Library Journals.js [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+		});
+	}
+	else {
+		invokeEMTranslator(doc, url);
+	}
+}
+
+function invokeEMTranslator(doc) {
+	var translator = Zotero.loadTranslator("web");
+	translator.setTranslator("951c027d-74ac-47d4-a107-9c3069ab7b48");
```
Comment (AbeJellinek) [inline Central and Eastern European Online Library Journals.js:72]:
```suggestion
	// Embedded Metadata
	translator.setTranslator("951c027d-74ac-47d4-a107-9c3069ab7b48");
```

File: Central and Eastern European Online Library Journals.js [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+	translator.setDocument(doc);
+	translator.setHandler("itemDone", function (t, i) {
+		postProcess(doc, i);
+	});
+	translator.translate();
+}
+
+//scraping abstractNote from HTML, that is not included in Embedded Metadata
```
Comment (AbeJellinek) [inline Central and Eastern European Online Library Journals.js:80]: formatting

File: Central and Eastern European Online Library Journals.js [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+function postProcess(doc, item) {
+	let abstractEntry = ZU.xpathText(doc, '//p[@class="summary"]');
+	if (!item.abstractNote && abstractEntry) item.abstractNote = abstractEntry;
+	item.complete();
```
Comment (AbeJellinek) [inline Central and Eastern European Online Library Journals.js:84]: Remove the Snapshot attachment - it's not needed for journal articles.

```js
for (let i = 0; i < item.attachments.length; i++) {
	if (item.attachments[i].title == 'Snapshot') {
		item.attachments.splice(i, 1);
		break;
	}
}
```

**Review by AbeJellinek — CHANGES_REQUESTED** [review]

File: Central and Eastern European Online Library Journals.js (hunk 3 above)
Comment (AbeJellinek) [reply Central and Eastern European Online Library Journals.js]: `a.ng-binding` will match every single tag on the page that AngularJS has bound data to. It's an internal bit of state and it isn't stable. `.description a` looks fine on my end. What do search results have to do with downloads?

File: Central and Eastern European Online Library Journals.js (hunk 5 above)
Comment (AbeJellinek) [reply Central and Eastern European Online Library Journals.js:313]: Add `"defer": true` if the page is populated by JavaScript.

File: Central and Eastern European Online Library Journals.js (hunk 3 above)
Comment (socheres) [reply Central and Eastern European Online Library Journals.js]: `.description a` looks fine in zotero client, but in Scaffold i got the error "detectWeb did not match". Anyway i changed it.

File: Central and Eastern European Online Library Journals.js (hunk 5 above)
Comment (socheres) [reply Central and Eastern European Online Library Journals.js:303]: How can i add `"defer": true`?

File: Central and Eastern European Online Library Journals.js (hunk 6 above)
Comment (socheres) [reply Central and Eastern European Online Library Journals.js:73]: ok

File: Central and Eastern European Online Library Journals.js (hunk 7 above)
Comment (socheres) [reply Central and Eastern European Online Library Journals.js:81]: what do you mean?

File: Central and Eastern European Online Library Journals.js (hunk 8 above)
Comment (socheres) [reply Central and Eastern European Online Library Journals.js:91]: Ok. I also had this in mind. Thanks!

File: Central and Eastern European Online Library Journals.js (hunk 5 above)
Comment (AbeJellinek) [reply Central and Eastern European Online Library Journals.js:303]: Like this:

```suggestion
		"defer": true,
		"items": [
```

File: Central and Eastern European Online Library Journals.js (hunk 7 above)
Comment (AbeJellinek) [reply Central and Eastern European Online Library Journals.js:81]: Sorry, should've been more specific. Take a look at the output that the linter gives when CI runs. In this case, you just need a space after the `//`.

File: Central and Eastern European Online Library Journals.js [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+	});
+	translator.translate();
+}
+
+//scraping abstractNote from HTML, that is not included in Embedded Metadata
+function postProcess(doc, item) {
+	let abstractEntry = ZU.xpathText(doc, '//p[@class="summary"]');
+	if (!item.abstractNote && abstractEntry) item.abstractNote = abstractEntry;
+	for (let i = 0; i < item.attachments.length; i++) {
+	if (item.attachments[i].title == 'Snapshot') {
```
Comment (AbeJellinek) [inline Central and Eastern European Online Library Journals.js:86]: Indent this!

Comment (socheres) [reply Central and Eastern European Online Library Journals.js]: lint check passed, but still got following error
```
TranslatorTester: Translating https://www.ceeol.com/search/journal-detail?id=1266
103
TranslatorTester: Central and Eastern European Online Library Journals Test 5: failed (Detection failed)
```

File: Central and Eastern European Online Library Journals.js (hunk 7 above)
Comment (socheres) [reply Central and Eastern European Online Library Journals.js]: ok

File: Central and Eastern European Online Library Journals.js (hunk 9 above)
Comment (socheres) [reply Central and Eastern European Online Library Journals.js]: thanks for your re-re-reviewing and your patience

File: Central and Eastern European Online Library Journals.js [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+			}
+		]
+	},
+	{
+		"type": "web",
+		"url": "https://www.ceeol.com/search/journal-detail?id=1266",
+		"defer": true,
+		"items": [
+			{
+				"itemType": "journalArticle",
```
Comment (AbeJellinek) [inline Central and Eastern European Online Library Journals.js:306]: This still isn't a journal article, right? It should be a multiple.

**Review by AbeJellinek — CHANGES_REQUESTED** [review]
Thanks for making those updates! Just the multiples stuff left.

File: Central and Eastern European Online Library Journals.js (hunk 3 above)
Comment (socheres) [reply Central and Eastern European Online Library Journals.js]: > `a.ng-binding` will match every single tag on the page that AngularJS has bound data to. It's an internal bit of state and it isn't stable. `.description a` looks fine on my end. What do search results have to do with downloads?

@AbeJellinek I have checked this again in Scaffold. With `.description a` it don't detect multiple select. On every journal page AngularJS bound the data, so from my point of view it is quite stable. If you know another css selector that would detect multiple select, just let me know.

File: Central and Eastern European Online Library Journals.js (hunk 10 above)
Comment (socheres) [reply Central and Eastern European Online Library Journals.js]: changed the css selector for detecting multiple downloads and updated the tests.

**Comment by AbeJellinek** [comment]
Merged in 9ba50a1276809f5c962e0d02618499cb8fcfad86. Thank you for all this great work!

**Pull request closed without merging**